File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if(len(prices) == 1):
            return 0

        maxProfit = 0

        left = 0
        right = 1

        while(right < len(prices)):
            if(prices[left] >= prices[right]):
                left = right
            else:
                maxProfit = max(maxProfit, prices[right]-prices[left])
        
            right += 1

        return maxProfit
        

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int[] maxPrice = new int[prices.length];
        int maxSoFar = 0;
        for(int i = prices.length-1; i >= 0; i--) {
            maxSoFar = Math.max(prices[i], maxSoFar);
            maxPrice[i] = maxSoFar;
        }

        int maxProfit = 0;
        for(int i = 0; i < prices.length; i++) {
            maxProfit = Math.max(maxProfit, maxPrice[i]-prices[i]);
        }

        return maxProfit;
    }
}